src/caf/toolkit/cost_utils.py

```diff
@@ -3,25 +3,416 @@
 from __future__ import annotations
 
 # Built-Ins
+import copy
 import logging
+import os
 
 from typing import Optional
+from typing import TYPE_CHECKING
 
 # Third Party
+import pydantic
 import numpy as np
+import pandas as pd
 
 # Local Imports
 # pylint: disable=import-error,wrong-import-position
+from caf.toolkit import math_utils
+from caf.toolkit import pandas_utils as pd_utils
 
 # pylint: enable=import-error,wrong-import-position
 
+if TYPE_CHECKING:
+    from dataclasses import dataclass
+else:
+    from pydantic.dataclasses import dataclass
+
 # # # CONSTANTS # # #
 LOG = logging.getLogger(__name__)
 
+
 # # # CLASSES # # #
+@dataclass(config={"arbitrary_types_allowed": True})  # type: ignore # pylint: disable=used-before-assignment, unexpected-keyword-arg
+class CostDistribution:
+    """Distribution of cost values between variable bounds.
+
+    Parameters
+    ----------
+    df:
+        The raw Pandas DataFrame containing the cost distribution data.
+
+    min_vals:
+        Minimum values of the cost distribution bin edges.
+
+    max_vals:
+        Maximum values of the cost distribution bin edges.
+
+    bin_edges:
+        Bin edges for the cost distribution.
+
+    avg_vals:
+        Average values for each of the cost distribution bins.
+
+    trip_vals:
+        Trip values for each of the cost distribution bins.
+
+    band_share_vals:
+        Band share values for each of the cost distribution bins.
+    """
+
+    df: pd.DataFrame
```

Comment:
How would that alias work? I tried having a search in the docs, but think I'm looking for the wrong words. If you know off the top of your head, that would help massively!

Once it's private, I'll just create a property for `df` so it can be gotten, but not changed at all.

Reply:
Pydantic has a Field class which can be used to define extra information about the attribute including alias, I think this should work `_df: pd.DataFrame = pydantic.Field(alias="df)`, the docs for field are https://docs.pydantic.dev/latest/usage/fields/

Reply:
This doesn't seem to play nice with the current code. Getting a validation error on constructor call when trying to access the `df`/`_df` item (I tired both just incase).
Is there no way to make `df` and init_only variable?

Using alias also produced PyLint errors everywhere as it doesn't recognise the different name.

Reply:
https://github.com/pydantic/pydantic/issues/4286 implies that this it to do with the pydantic.dataclass and alias not playing nice together. Only fixed in V2, which breaks the rest of the code

Reply:
@wsp-mbuckley, are you happy to approve these changes? 

I'll make an issue for this one too, then we can come back to it.

Reply:
Issue added at #65